backend/proxy/proxy.go

```
@@ -0,0 +1,78 @@
+// Copyright 2019 Shift Devices AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package socksproxy
+
+import (
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+
+	"golang.org/x/net/proxy"
+)
+
+// SocksProxy holds the proxy address and wether to use it
+type SocksProxy struct {
+	useProxy     bool
+	proxyAddress string
+}
+
+// NewSocksProxy returns a new socks proxy instance
+func NewSocksProxy(useProxy bool, proxyAddress string) SocksProxy {
+	proxy := SocksProxy{
+		useProxy:     useProxy,
+		proxyAddress: proxyAddress,
+	}
+	return proxy
+}
+
+// GetTCPConnection returns a tcp connection. The connection is proxied, if useProxy is true.
+func (socksProxy *SocksProxy) GetTCPConnection(address string) (net.Conn, error) {
+	if socksProxy.useProxy {
+		// Create a proxy that uses Tor's SocksPort.
+		dialer, err := proxy.SOCKS5("tcp", socksProxy.proxyAddress, nil, nil)
+		if err != nil {
+			log.Fatal(err)
+		}
+		return dialer.Dial("tcp", address)
+	}
+	return net.Dial("tcp", address)
+}
+
+// GetHTTPClient returns a http client. Requests made with this client are proxied, if useProxy is true.
+func (socksProxy *SocksProxy) GetHTTPClient() *http.Client {
```

the error branches should probably return the error